src/ipc/ipc4/handler.c

```diff
@@ -828,6 +828,99 @@ static int ipc4_get_large_config_module_instance(struct ipc4_message_request *ip
 	return ret;
 }
 
+static int ipc4_set_vendor_config_module_instance(uint32_t module_id,
+						  uint32_t instance_id,
+						  bool init_block,
+						  bool final_block,
+						  uint32_t data_off_size,
+						  const char *data)
+{
+	struct comp_dev *dev = NULL;
+	int ret;
+	const struct comp_driver *drv = ipc4_get_comp_drv(module_id);
+
+	if (module_id) {
+		uint32_t comp_id;
+
+		comp_id = IPC4_COMP_ID(module_id, instance_id);
+		dev = ipc4_get_comp_dev(comp_id);
+		if (!dev)
+			return IPC4_MOD_INVALID_ID;
+
+		/* Pass IPC to target core */
+		if (!cpu_is_me(dev->ipc_config.core))
+			return ipc4_process_on_core(dev->ipc_config.core, false);
+	}
+
+	if (init_block && final_block) { // bursted configs
+		const struct ipc4_mod_conf_param *tlv = (struct ipc4_mod_conf_param *)data;
+		/* if there is no payload in this large config set
+		 * (4 bytes param_id | 4 bytes param_size=0 | no param_data)
+		 * then we fail
+		 */
+		if (data_off_size < sizeof(struct ipc4_mod_conf_param))
+			return IPC4_ERROR_INVALID_PARAM;
+
+		/* ===Iterate over payload===
+		 * Payload can have multiple ipc4_mod_conf_param structures inside,
+		 * You can find how many by checking payload size (data_off_size)
+		 * Here we just set pointer end_offset to the end of data
+		 * and iterate until we reach that
+		 */
+
+		const uint8_t *end_offset = data + data_off_size;
+
+		if ((const uint8_t *)tlv >= end_offset) {
+			/* This means that there is indeed nothing to be done here */
+			return IPC4_INVALID_CONFIG_DATA_STRUCT;
+		}
+
+		while ((const uint8_t *)tlv < end_offset) {
+			const struct byte_array_simple array = {
+					(uint8_t *)&tlv->param_data[0], tlv->param_size };
+
+			ret = drv->ops.set_large_config(dev, tlv->param_id, init_block,
```

Comment:
Good catch. From what I can see here, set_large_config implementations do not check for `dev == NULL` either , and first thing those do is trying to access  `dev->priv_data`.

The problem is, I based this function on ipc4_get_large_config_module_instance which does the same thing. Here we also write to dev variable only if module_id is not 0 (because base_fw has no dev apparently). 

After searching a bit I saw that base_fw also implements large config set in `basefw_set_large_config` function. It takes comp_dev argument just to be compatible with the interface, but assumes it is NULL as it is not used. 

So, it seems it really can be called with `dev == NULL` in this one case.